### PR TITLE
Handle recursion in `clone`

### LINKED ISSUE #9795: Array#clone, and probably others, can't handle cycles in data.
Array#clone is non-terminatingly recursive:
```crystal
alias X = Array(X)|Int32
a = Array(X).new
a << a
b = a.clone
```
One fix would be to document the behavior that it can't handle cycles in data, and leave it at that.
Another would be to add a loop detector which would properly reference an already-copied data structure rather than cloning it again. This would have a slight speed and complexity penalty.

Yes, I know that cycles are bad for garbage collection. The particular application I am working on goes to some pains to recursively clear its used data (with a loop detector), because it's real-time and shouldn't ever be stopped.

### PR DESCRIPTION
**Pull request by asterite**
Fixes #9795

Note that we do detect recursion on other methods (like `inspect`) so I think this makes sense as a change.

This handles recursion in `clone` in these types:
- Array
- Hash
- Deque
- And type that uses `def_clone`

I forgot to mention: the recursive check is not done in a few cases, like when it's an array of integers, bools or string, where recursion can't happen. Ideally we'd have a way to detect that a type doesn't have pointers, but it's not simple.

For now, it's better this works correctly. We can optimize it later on. In my benchmarks, the new `clone` takes twice the time, but it's still fast.

### CODE REVIEW

File: src/hash.cr [hunk 1]
```diff
@@ -1801,9 +1801,18 @@ class Hash(K, V)
   # hash_a # => {"foobar" => {"foo" => "bar"}}
   # ```
   def clone
-    hash = Hash(K, V).new
-    hash.initialize_clone(self)
-    hash
+    {% if V == ::Bool || V == ::Char || V == ::String || V == ::Symbol || V < ::Number::Primitive %}
```
Comment (bcardiff) [inline src/hash.cr:1804]: Shouldn't `K` be mentioned here also?

**Review by bcardiff — COMMENTED** [review]
Hashes with recursive keys are not handled well it seems.

```crystal
alias H = Hash(H, H)

h = Hash(H, H).new
h[h] = h

pp! h # => {{...} => {...}}
h.each do |k, v|
  pp! k.object_id == h.object_id # => true
  pp! v.object_id == h.object_id # => true
end

h2 = h.clone

pp! h2 # => {{{...} => {...}} => {...}}  # there is something different in h2 w.r.t. h
h2.each do |k, v|
  pp! k.object_id == h2.object_id # => false # this should be true
  pp! v.object_id == h2.object_id # => true
end
```
---

I wonder if a flag like `compare_by_identity` or argument in clone should be added to allow users to do a faster clone as a workaround...

**Comment by asterite** [comment]
> Hashes with recursive keys are not handled well it seems

It's on purpose. Clone doesn't clone the keys. That's how it was and I didn't want to change that.

However, if you want we can also change clone to clone the keys.

File: src/hash.cr (hunk 1 above)
Comment (asterite) [reply src/hash.cr:1804]: No because keys are not cloned.

**Comment by bcardiff** [comment]
Since cloning should preserve == and hash I see little reason to keep the orginal keys instead of clonning them. Unless the hash was with `compare_by_identity` 🤔 .

It seems ruby (2.3.7) does not behave correctly in this case either. It seems to not even clone the value correctly

```ruby
a = {}
a[a] = a 
a # => {{...}=>{...}}
b = a.clone
b # => {{{...}=>{...}}=>{{...}=>{...}}}

b.keys.first == b # => false
b.keys.first == a # => true

b.values.first == b # => false
b.values.first == a # => true
```

I'm uncertain.

**Comment by asterite** [comment]
That's because `clone` in Ruby doesn't do a deep clone. Such concept doesn't exist. That's why people use Marshal.dump and load to clone objects, there's no other way.

More info: https://ruby-doc.org/core-2.7.1/Object.html#method-i-dup-label-on+dup+vs+clone

I think it's fine if we clone the key, I don't know why it's not like that right now. Let me know and I can push one more commit.

**Comment by asterite** [comment]
It seems I introduced `Hash#clone` in 2013: https://github.com/crystal-lang/crystal/commit/8ef0967f98 (WAAAAAAT)

I don't know why I didn't clone the key.

**Comment by bcardiff** [comment]
I think keys should be cloned, but it can be handled in another PR if wanted. Usually hashes are used with simple types.

It needs a rebase on master due to latest changes though.

**Review by bcardiff — APPROVED** [review]

**Comment by asterite** [comment]
> It needs a rebase on master due to latest changes though

The branch is in this repo, next time feel free to rebase it and continue it :-)

**Pull request merged**